larpandoracontent/LArObjects/LArAdaBoostDecisionTree.cc

```diff
@@ -0,0 +1,394 @@
+    }
+
+    if (!pContainerXmlNode)
+    {
+        std::cout << "AdaBoostDecisionTree: Could not find an AdaBoostDecisionTree of name " << bdtName << std::endl;
+        return pandora::STATUS_CODE_NOT_FOUND;
+    }
+
+    pandora::TiXmlHandle localHandle(pContainerXmlNode);
+    pandora::TiXmlElement *pCurrentXmlElement = localHandle.FirstChild().Element();
+
+    while (pCurrentXmlElement)
+    {
+        if (pandora::STATUS_CODE_SUCCESS != this->ReadComponent(pCurrentXmlElement, weakClassifiers))
+        {
+            std::cout << "AdaBoostDecisionTree: Unknown component in xml file" << std::endl;
+            return pandora::STATUS_CODE_FAILURE;
+        }
+
+        pCurrentXmlElement = pCurrentXmlElement->NextSiblingElement();
+    }
+    return pandora::STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+pandora::StatusCode AdaBoostDecisionTree::ReadComponent(pandora::TiXmlElement *pCurrentXmlElement, WeakClassifiers &weakClassifiers) const 
+{
+    const std::string componentName(pCurrentXmlElement->ValueStr());
+    pandora::TiXmlHandle currentHandle(pCurrentXmlElement);
+
+    if ((std::string("Name") == componentName) || (std::string("Timestamp") == componentName))
+        return pandora::STATUS_CODE_SUCCESS;
+
+    if (std::string("DecisionTree") == componentName)
+        return this->ReadDecisionTree(currentHandle, weakClassifiers);
+
+    return pandora::STATUS_CODE_INVALID_PARAMETER;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+pandora::StatusCode AdaBoostDecisionTree::ReadDecisionTree(pandora::TiXmlHandle &currentHandle, WeakClassifiers &weakClassifiers) const
+{
+    IdToNodeMap idToNodeMap;
+    double boostWeight(0.);
+    int treeIndex(0);
+
+    for (pandora::TiXmlElement *pHeadTiXmlElement = currentHandle.FirstChildElement().ToElement(); pHeadTiXmlElement != NULL; pHeadTiXmlElement = pHeadTiXmlElement->NextSiblingElement())
+    {
+        if ("TreeIndex" == pHeadTiXmlElement->ValueStr())
+        {
+            PANDORA_RETURN_RESULT_IF_AND_IF(pandora::STATUS_CODE_SUCCESS, pandora::STATUS_CODE_NOT_FOUND, !=, pandora::XmlHelper::ReadValue(currentHandle,
+                "TreeIndex", treeIndex));
+        }
+        else if ("BoostWeight" == pHeadTiXmlElement->ValueStr())
+        {
+            PANDORA_RETURN_RESULT_IF_AND_IF(pandora::STATUS_CODE_SUCCESS, pandora::STATUS_CODE_NOT_FOUND, !=, pandora::XmlHelper::ReadValue(currentHandle,
+                "BoostWeight", boostWeight));
+        }
+        else if ("Node" == pHeadTiXmlElement->ValueStr())
+        {
+            pandora::TiXmlHandle nodeHandle(pHeadTiXmlElement);
+            this->ReadNode(nodeHandle, idToNodeMap);
+        }
+    }
+
+    const WeakClassifier *pWeakClassifiers = new WeakClassifier(idToNodeMap, boostWeight, treeIndex);
```

Comment:
We should talk about this tomorrow. Still heap allocating weak classifiers in the decision tree class, rather than hiding all the weak classifier knowledge inside the strong classifier.

Can you just give the strong classifier (and then weak classifier) constructors that take xmlHandles?

The diff is getting hard to read, but it looks like the following should work:

-The decision tree should new up a strong classifier, and delete it in its destructor (and make sure that it doesn't leak the memory if the job terminates early).

-The strong classifier should new up its weak classifiers, and delete them in its destructor (and make sure that it doesn't leak memory if exceptions cause the job to terminate early).

-The weak classifiers should new up its nodes, and delete them in its destructor, and so on.

In this way, each object is encapsulated and there isn't a decision tree object that can mess about with the contents of each.